Tasks/AzureRmWebAppDeployment/KuduUtility.ts

```diff
@@ -0,0 +1,89 @@
+import Q = require('q');
+import tl = require('vsts-task-lib/task');
+import path = require("path");
+import fs = require("fs");
+var request = require ('request');
+var archiver = require('archiver');
+
+
+/**
+ *  Deploys a zip based webapp package.
+ * 
+ *  @param  webAppPackage                  Zip file or folder for deployment
+ *  @param  virtualApplicationMappings     Mapping to get physical path for deployment
+ *  @param  publishingProfile              publish profile provides destination details for deployment
+ *  @param  virtualApplication             (Optional) Virtual application name 
+ */
+export async function deployWebAppPackage(webAppPackage: string, virtualApplicationMappings, publishingProfile,
+                             virtualApplication: string) {
+
+    var deferred = Q.defer<any>();	
+    // construct URL depending on virtualApplication or root of webapplication 
+    var physicalDeploymentPath = "/site/wwwroot";
+    var virtualPath = "/";
+    if (virtualApplication) {
+		virtualPath = "/" + virtualApplication;
+    }
+
+    for( var index in virtualApplicationMappings ){
+        var mapping = virtualApplicationMappings[index];
+        if( mapping.virtualPath == virtualPath){
+            physicalDeploymentPath = mapping.physicalPath;
+            break;
+        }
+    }
+	
+    tl.debug(tl.loc("Deployingwebapplicationatvirtualpathandphysicalpath", webAppPackage, virtualPath, physicalDeploymentPath));
+
+    var kuduDeploymentURL = "https://" + publishingProfile.publishUrl + "/api/zip/" + physicalDeploymentPath;
+    var basicAuthToken = 'Basic ' + new Buffer(publishingProfile.userName + ':' + publishingProfile.userPWD).toString('base64');
+    await fs.createReadStream(webAppPackage).pipe(request.put({ url: kuduDeploymentURL, headers: {
+            'Authorization': basicAuthToken,
+            'content-type': 'multipart/form-data'
+        } }, function (error, response) {
+            if (error){
+                deferred.reject(tl.loc("Failedtodeploywebapppackageusingkuduservice",error));
+            } else if(response.statusCode === 200) {
+                tl._writeLine(tl.loc("Successfullydeployedpackageusingkuduserviceat", webAppPackage, publishingProfile.publishUrl));
+                deferred.resolve(tl.loc("Successfullydeployedpackageusingkuduserviceat", webAppPackage, publishingProfile.publishUrl))
+            }
+            else {
+                tl.debug("Response :"+ JSON.stringify(response));
+                deferred.reject(tl.loc('Unabletodeploywebappresponsecode', response.statusCode));
+            }
+    }));
+    return deferred.promise;
+}
+
+function onError(error) {
+    tl.setResult(tl.TaskResult.Failed, error);
+    process.exit(1);
```

Comment:
Never process.exit from a task.  We had edge cases where some output was lost and it was the last line written which is often critical - fail the task.  Exactly what you have here cost us quite awhile to figure out with minecraft.  1 in so many 100s of builds would lose last line on some agents which is why we deprecated tl.exit().  You need to make the flow of the task  handle failures.  If you're in promise chains or an async function you can throw.